### PR TITLE
Server pdf format options

### PR DESCRIPTION
**Pull request by mhugent**
In QGIS Server, there is the FORMAT_OPTIONS parameter to pass format options for DXF format to WMS GetPrint. This PR adds support for format options for PDF format in WMS GetPrint

### CODE REVIEW

File: src/server/services/wms/qgswmsparameters.cpp [hunk 1]
```diff
@@ -2210,35 +2210,156 @@ namespace QgsWms
...
+    return useOgcGeoreferencing;
+  }
+
+  QStringList QgsWmsParameters::pdfExportMapThemes() const
+  {
+    QStringList themes;
+    const QMap<QgsWmsParameters::PdfFormatOption, QString> options = formatOptions<QgsWmsParameters::PdfFormatOption>();
+    if ( options.contains( PdfFormatOption::EXPORT_THEMES ) )
+    {
+      themes = options[PdfFormatOption::EXPORT_THEMES].split( "," );
```
Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.cpp:2340]:
```suggestion
      themes = options[PdfFormatOption::EXPORT_THEMES].split( ',' );
```

File: src/server/services/wms/qgswmsparameters.cpp [hunk 2]
```diff
@@ -2210,35 +2210,156 @@ namespace QgsWms
...
+  }
 
-    return options;
+  QVector<qreal> QgsWmsParameters::pdfPredefinedMapScales() const
+  {
+    QVector<qreal> scales;
+    const QMap<QgsWmsParameters::PdfFormatOption, QString> options = formatOptions<QgsWmsParameters::PdfFormatOption>();
+    if ( options.contains( PdfFormatOption::PREDEFINED_MAP_SCALES ) )
+    {
+      QStringList scaleList = options[PdfFormatOption::PREDEFINED_MAP_SCALES].split( "," );
```
Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.cpp:2351]:
```suggestion
      QStringList scaleList = options[PdfFormatOption::PREDEFINED_MAP_SCALES].split( ',' );
```

Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.cpp:2351]:
```suggestion
      const QStringList scaleList = options[PdfFormatOption::PREDEFINED_MAP_SCALES].split( "," );
```

File: src/server/services/wms/qgswmsparameters.cpp [hunk 3]
```diff
@@ -2210,35 +2210,156 @@ namespace QgsWms
...
+      for ( const auto &it : scaleList )
```
Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.cpp:2352]:
```suggestion
      for ( const QString &it : std::as_const( scaleList ) )
```

File: src/server/services/wms/qgswmsparameters.cpp [hunk 4]
```diff
@@ -2210,35 +2210,156 @@ namespace QgsWms
...
+    const QMap<QgsWmsParameters::PdfFormatOption, QString> options = formatOptions<QgsWmsParameters::PdfFormatOption>();
+    if ( options.contains( PdfFormatOption::EXPORT_THEMES ) )
+    {
+      themes = options[PdfFormatOption::EXPORT_THEMES].split( "," );
+    }
+    return themes;
+  }
 
-    return options;
+  QVector<qreal> QgsWmsParameters::pdfPredefinedMapScales() const
```
Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.cpp:2345]: Why `qreal` and not `double`?

File: src/server/services/wms/qgswmsparameters.h [hunk 5]
```diff
@@ -1367,11 +1380,114 @@ namespace QgsWms
...
+      bool pdfExportMetadata() const;
+
+      /**
+       * Returns text render format for pdf export
+       * \since QGIS 3.32
+       */
+      Qgis::TextRenderFormat pdfTextRenderFormat() const;
+
+      /**
+       * Returns true if images embedded in pdf must be lossless
```
Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.h:1420]:
```suggestion
       * Returns true if images embedded in pdf must be compressed using a lossless algorithm
```

File: src/server/services/wms/qgswmsparameters.h [hunk 6]
```diff
@@ -1367,11 +1380,114 @@ namespace QgsWms
...
+       * Returns map themes for GeoPDF export
+       * \since QGIS 3.32
+       */
+      QStringList pdfExportMapThemes() const;
+
+      /**
+       * Returns list of map scales
+       * \since QGIS 3.32
+       */
+      QVector<qreal> pdfPredefinedMapScales() const;
```
Comment (elpaso) [inline src/server/services/wms/qgswmsparameters.h:1453]: double ?

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 7]
```diff
@@ -630,8 +630,37 @@ namespace QgsWms
       exportSettings.flags |= QgsLayoutRenderContext::FlagDrawSelection;
       // Print as raster
       exportSettings.rasterizeWholeImage = layout->customProperty( QStringLiteral( "rasterize" ), false ).toBool();
-      // Set scales
-      exportSettings.predefinedMapScales = QgsLayoutUtils::predefinedScales( layout.get( ) );
+      // Set scales. 1. Prio: request, 2. Prio: predefined mapscales in layout
+      QVector<qreal> requestMapScales = mWmsParameters.pdfPredefinedMapScales();
```
Comment (elpaso) [inline src/server/services/wms/qgswmsrenderer.cpp:634]: double?

**Review by elpaso — COMMENTED** [review]
LGTM, just left some minor comments

**Comment by mhugent** [comment]
Thanks for reviewing @elpaso

**Review by manisandro — APPROVED** [review]
LGTM

**Comment by DelazJ** [comment]
> In QGIS Server, there is the FORMAT_OPTIONS parameter to pass format options for DXF format to WMS GetPrint.

Browsing the docs, FORMAT_OPTIONS (for Dxf) is mentioned as parameter only in [WMS GetMap request](https://docs.qgis.org/3.28/fr/docs/server_manual/services/wms.html?highlight=format_options#getmap). No GetPrint. @mhugent this is a typo, right?
Moreover, I think the changelog is totally wrong (or I totally misunderstand what this PR is about), [when stating](https://qgis.org/fr/site/forusers/visualchangelog332/index.html#feature-pdf-format-options-for-wms-getprint):
> In QGIS Server, there is now a FORMAT_OPTIONS parameter which can pass format options for DXF format to WMS GetPrint. This allows more control in GetPrint requests when generating GeoPDF outputs, such as Georeference or Geometry Simplification settings.
>
>This feature was developed by [mhugent](https://github.com/mhugent)

Any clarification, please? Thanks.

**Comment by DelazJ** [comment]
I try to document this feature as I understood it at https://github.com/qgis/QGIS-Documentation/pull/8379. Feedback would be very much appreciated. Thanks. (and it would be good to know whether the changelog is misleading)

**Comment by mhugent** [comment]
Before the PR, the FORMAT_OPTIONS parameter was only for dxf. And dxf is only supported for GetMap, so the statement was true. But now, it can also be used for pdf. And pdf is only supported as output format in GetPrint. So now, FORMAT_OPTIONS can be used in GetMap for dxf and GetPrint for pdf. 
The changelog should be like this:
The FORMAT_OPTIONS parameter in QGIS server can now also be used to pass format options for PDF format to WMS GetPrint. This allows more control in GetPrint requests when generating GeoPDF outputs, such as Georeference or Geometry Simplification settings.

**Comment by DelazJ** [comment]
Thanks for confirmation @mhugent. Changelog fix is underway.

**Pull request merged**